File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    min_time = 1
    max_time = min(times) * n
    mid_time = max_time // 2

    while min_time < max_time:
        possible_number = 0
        for t in times:
            possible_number += mid_time // t

        if possible_number >= n:
            max_time = mid_time
        else:
            min_time = mid_time + 1
        mid_time = min_time + (max_time - min_time) // 2

    return mid_time

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.Arrays;

class Solution {
	public long solution(int n, int[] times) {
		Arrays.sort(times);
		long loTime = 0;
		long hiTime = (long) times[0] * n;

		while (loTime < hiTime) {
			long midTime = loTime + (hiTime - loTime) / 2;

			long peopleCnt = 0;
			for (int waitTime : times) {
				peopleCnt += midTime / waitTime;
				if (peopleCnt >= n)
					break;
			}

			if (peopleCnt >= n) {
				hiTime = midTime;
			} else {
				loTime = midTime + 1;
			}
		}

		return loTime;
	}
}
